app/src/org/commcare/tasks/DataPullTask.java

```diff
@@ -104,249 +111,291 @@ private DataPullTask(String username, String password, String server, Context co
+
+            @Override
+            public void reportProgress(int progress) {
+                if (!publishedAuth) {
+                    DataPullTask.this.publishProgress(PROGRESS_AUTHED, progress);
+                    publishedAuth = true;
+                }
+            }
+        };
+        Logger.log(AndroidLogger.TYPE_USER, "Starting Sync");
+
+        // This is a dangerous way to do this (the null settings), should revisit later
+        if (loginNeeded) {
+            if (!useExternalKeys) {
+                // Get the key
+                SecretKey newKey = CryptUtil.generateSemiRandomKey();
+
+                if (newKey == null) {
+                    this.publishProgress(PROGRESS_DONE);
+                    responseError = PullTaskResult.UNKNOWN_FAILURE;
+                    return;
+                }
+                String sandboxId = PropertyUtils.genUUID().replace("-", "");
+                ukr = new UserKeyRecord(username, UserKeyRecord.generatePwdHash(password),
+                        CryptUtil.wrapByteArrayWithString(newKey.getEncoded(), password),
+                        new Date(), new Date(Long.MAX_VALUE), sandboxId);
 
-            //Whether or not we should be generating the first key
-            boolean useExternalKeys = !(keyServer == null || keyServer.equals(""));
+            } else {
+                ukr = UserKeyRecord.getCurrentValidRecordByPassword(app, username, password, true);
+                if (ukr == null) {
+                    Logger.log(AndroidLogger.TYPE_ERROR_ASSERTION, "Shouldn't be able to not have a valid key record when OTA restoring with a key server");
+                    this.publishProgress(PROGRESS_DONE);
+                    responseError = PullTaskResult.UNKNOWN_FAILURE;
+                    return;
+                }
+            }
 
-            boolean loginNeeded = true;
-            boolean useRequestFlags = false;
+            //add to transaction parser factory
+            byte[] wrappedKey = CryptUtil.wrapByteArrayWithString(ukr.getEncryptedKey(), password);
+            factory.initUserParser(wrappedKey);
+        } else {
             try {
-                loginNeeded = !CommCareApplication._().getSession().isActive();
+                factory.initUserParser(CommCareApplication._().getSession().getLoggedInUser().getWrappedKey());
             } catch (SessionUnavailableException sue) {
-                //expected if we aren't initialized.
+                // TODO PLM: eventually take out this catch. These should be
+                // checked locally
+                //TODO: Keys were lost somehow.
+                sue.printStackTrace();
+                responseError = PullTaskResult.UNKNOWN_FAILURE;
+                return;
             }
 
-            PullTaskResult responseError = PullTaskResult.UNKNOWN_FAILURE;
+            // Only purge cases if we already had a logged in user. Otherwise we probably can't read the DB.
+            CaseUtils.purgeCases();

```

Comment:
Wondering if we can move this out of the `setup` method, because at this point the cancel button is being shown, but won't dismiss until this finishes, so there may be some lag for the user if they press the cancel button and there are a lot of cases to purge